spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/alloydb/AlloyDbEnvironmentPostProcessor.java

```diff
@@ -0,0 +1,155 @@
+    AlloyDbProperties alloyDbProperties = new Binder(
+        ConfigurationPropertySources.get(environment),
+        new NonSecretsManagerPropertiesPlaceholdersResolver(environment),
+        null,
+        null,
+        null)
+        .bind(propertiesPrefix, AlloyDbProperties.class)
+        .orElse(new AlloyDbProperties());
+
+    if (isOnClasspath("com.google.cloud.alloydb.SocketFactory")
+        && isOnClasspath(JDBC_DRIVER_CLASS)) {
+      // configure default JDBC driver and username as fallback values when not
+      // specified
+      Map<String, Object> fallbackMap = new HashMap<>();
+      fallbackMap.put("spring.datasource.username", DEFAULT_USERNAME);
+      fallbackMap.put(
+          "spring.datasource.driver-class-name", JDBC_DRIVER_CLASS);
+      environment
+          .getPropertySources()
+          .addLast(new MapPropertySource("ALLOYDB_DATA_SOURCE_FALLBACK", fallbackMap));
+
+      // always set the spring.datasource.url property in the environment
+      Map<String, Object> primaryMap = new HashMap<>();
+      primaryMap.put("spring.datasource.url", getJdbcUrl(alloyDbProperties));
+      environment
+          .getPropertySources()
+          .addFirst(new MapPropertySource("ALLOYDB_DATA_SOURCE_URL", primaryMap));
+    }
+
+    // support usage metrics
+    ConnectorRegistry.addArtifactId(
+        "spring-cloud-gcp-alloydb/" + this.getClass().getPackage().getImplementationVersion());
+  }
+
+  private String getJdbcUrl(AlloyDbProperties properties) {
+    String jdbcUrl = String.format(JDBC_URL_TEMPLATE, properties.getDatabaseName());
+
+    if (StringUtils.hasText(properties.getInstanceConnectionUri())) {
+      jdbcUrl += "&alloydbInstanceName=" + properties.getInstanceConnectionUri();
+    }
+
+    if (StringUtils.hasText(properties.getIpType())) {
+      jdbcUrl += "&alloydbIpType=" + properties.getIpType();
+    }
+
+    if (properties.isEnableIamAuth()) {
+      jdbcUrl += "&alloydbEnableIAMAuth=true&sslmode=disable";
+    }
+
+    if (StringUtils.hasText(properties.getAdminServiceEndpoint())) {
+      jdbcUrl += "&alloydbAdminServiceEndpoint=" + properties.getAdminServiceEndpoint();
+    }
+
+    if (StringUtils.hasText(properties.getQuotaProject())) {
+      jdbcUrl += "&alloydbQuotaProject=" + properties.getQuotaProject();
+    }
+
+    if (StringUtils.hasText(properties.getTargetPrincipal())) {
+      jdbcUrl += "&alloydbTargetPrincipal=" + properties.getTargetPrincipal();
+    }
+
+    if (StringUtils.hasText(properties.getDelegates())) {
+      jdbcUrl += "&alloydbDelegates=" + properties.getDelegates();
+    }
+
+    if (StringUtils.hasText(properties.getNamedConnector())) {
+      jdbcUrl += "&alloydbNamedConnector=" + properties.getNamedConnector();
+    }
+
+    return jdbcUrl;
+  }
+
+  private boolean isOnClasspath(String className) {
+    return ClassUtils.isPresent(className, null);
+  }
+
+  private static class NonSecretsManagerPropertiesPlaceholdersResolver
```

Comment:
What is the purpose of this resolver? I don't think we use anything like this in the `CloudSqlEnvironmentPostProcessor`.

Reply:
It is present in [PropertiesRetriever.java](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/main/spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/sql/PropertiesRetriever.java#L60). It was introduced by https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/495 and moved to `PropertiesRetriever` by https://github.com/GoogleCloudPlatform/spring-cloud-gcp/pull/772.

Please let me know if it is not required for AlloyDB.